src/plugins/procmon/winnt.cpp

```diff
@@ -0,0 +1,174 @@
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "winnt.h"
+#include <stdio.h>
+
+std::string stringify_protection_attributes(uint32_t attributes, char sep)
+{
+    if (attributes == 0)
+    {
+        return std::string("NONE");
+    }
+
+    std::string result;
+
+    int num = 0;
+    for (size_t i = 0; i < 8 * sizeof(attributes) && attributes; ++i)
+    {
+        uint32_t attr = attributes & (1 << i);
+        if (attr == 0)
+            continue;
+
+        attributes &= ~attr;
+        ++num;
+        if (num > 1)
+            result.append(1, sep);
+
+        switch (attr)
+        {
+            case PAGE_NOACCESS:
+                result.append("PAGE_NOACCESS");
+                break;
+            case PAGE_READONLY:
+                result.append("PAGE_READONLY");
+                break;
+            case PAGE_READWRITE:
+                result.append("PAGE_READWRITE");
+                break;
+            case PAGE_WRITECOPY:
+                result.append("PAGE_WRITECOPY");
+                break;
+            case PAGE_EXECUTE:
+                result.append("PAGE_EXECUTE");
+                break;
+            case PAGE_EXECUTE_READ:
+                result.append("PAGE_EXECUTE_READ");
+                break;
+            case PAGE_EXECUTE_READWRITE:
+                result.append("PAGE_EXECUTE_READWRITE");
+                break;
+            case PAGE_EXECUTE_WRITECOPY:
+                result.append("PAGE_EXECUTE_WRITECOPY");
+                break;
+            case PAGE_GUARD:
+                result.append("PAGE_GUARD");
+                break;
+            case PAGE_NOCACHE:
+                result.append("PAGE_NOCACHE");
+                break;
+            case PAGE_WRITECOMBINE:
+                result.append("PAGE_WRITECOMBINE");
+                break;
+            default:
+            {
+                char tmp[32] = {0};
+                sprintf(tmp, "0x%" PRIx32, attr);
+                result.append(tmp);
+                break;
+            }
+        }
+    }
+    return result;
```

Comment:
Why not return `result.c_str()`?

Reply:
string object will be destroyed, so c_str() returns invalid memory pointer

Reply:
I see